Sources/CAtomics/include/CAtomics.h

```diff
@@ -535,3 +535,103 @@ _Bool UnmanagedCompareAndSwap(OpaqueUnmanagedHelper *_Nonnull ptr, const void *_
 
 #undef __CACHE_LINE_WIDTH
 #endif
+
+// tagged
+
+CLANG_ATOMICS_STRUCT(TaggedOpaqueUnmanagedHelper, _Atomic(__UNION_TYPE), _Alignof(_Atomic(__UNION_TYPE)))
+CLANG_ATOMICS_IS_LOCK_FREE(TaggedOpaqueUnmanagedHelper)
+CLANG_ATOMICS_TAGGED_POINTER_INITIALIZE(TaggedOpaqueUnmanagedHelper, TaggedOptionalRawPointer)
+CLANG_ATOMICS_TAGGED_POINTER_LOAD(TaggedOpaqueUnmanagedHelper, TaggedOptionalRawPointer)
+CLANG_ATOMICS_TAGGED_POINTER_STORE(TaggedOpaqueUnmanagedHelper, TaggedOptionalRawPointer)
+
+static __inline__ __attribute__((__always_inline__)) \
+SWIFT_NAME(TaggedOpaqueUnmanagedHelper.lockAndLoad(self:_:)) \
+TaggedOptionalRawPointer TaggedUnmanagedLockAndLoad(TaggedOpaqueUnmanagedHelper *_Nonnull ptr, enum LoadMemoryOrder order)
+{ // load the pointer value, and leave the pointer either LOCKED or NULL; spin for the lock if necessary
+#ifndef __SSE2__
+    char c;
+    c = 0;
+#endif
+    TaggedOptionalRawPointer rp;
+    rp.tag_ptr = atomic_load_explicit(&(ptr->a), order);
+    do { // don't fruitlessly invalidate the cache line if the value is locked
+        while (rp.tag_ptr == __OPAQUE_UNMANAGED_LOCKED)
```

Comment:
I would feel better if this comparison was with a constant of the same type. You could define a `__TAGGED_UNMANAGED_LOCKED` of type __UNION_TYPE and appropriate value. The 0x7 should be in the pointer field and the tag should be 0. Otherwise we could find ourselves confusing it with a NULL pointer with a tag of 7.